File: mirea/cources/2026-2027/aii-july/06/Part-04-Meet-in-the-middle/examples/example-02.py
# Пример: максимальная сумма подмножества, не превосходящая limit_sum.
# Решаем ту же задачу через meet-in-the-middle.

from bisect import bisect_right


# Строим все суммы подмножеств для одной половины массива.
def build_sums(part: list[int]) -> list[int]:
	sums = []
	for mask in range(1 << len(part)):
		current_sum = 0
		for i in range(len(part)):
			if (mask >> i) & 1:
				current_sum += part[i]
		sums.append(current_sum)
	return sums


n, limit_sum = map(int, input().split())
a = list(map(int, input().split()))

# Делим массив на две половины, чтобы перебирать их независимо.
mid = n // 2
left = a[:mid]
right = a[mid:]

# Генерируем все суммы отдельно для каждой половины.
left_sums = build_sums(left)
right_sums = sorted(build_sums(right))

best = 0
for left_sum in left_sums:
	if left_sum > limit_sum:
		continue

	# Ищем максимальную допустимую сумму правой половины.
	need = limit_sum - left_sum
	pos = bisect_right(right_sums, need) - 1
	if pos >= 0:
		best = max(best, left_sum + right_sums[pos])
	else:
		best = max(best, left_sum)

# Печатаем лучшую допустимую сумму.
print(best)

File: mirea/cources/2026-2027/aii-july/06/Part-04-Meet-in-the-middle/examples/example-02.cpp
// Пример: максимальная сумма подмножества, не превосходящая limit_sum.
// Решаем ту же задачу через meet-in-the-middle.

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

// Строим все суммы подмножеств для одной половины массива.
vector<long long> build_sums(const vector<long long>& part) {
    vector<long long> sums;
    int m = (int)part.size();
    sums.reserve(1 << m);
    for (int mask = 0; mask < (1 << m); mask++) {
        long long current_sum = 0;
        for (int i = 0; i < m; i++) {
            if ((mask >> i) & 1) current_sum += part[i];
        }
        sums.push_back(current_sum);
    }
    return sums;
}

int main() {
    int n;
    long long limit_sum;
    // Считываем массив и ограничение на сумму.
    cin >> n >> limit_sum;

    vector<long long> a(n);
    for (int i = 0; i < n; i++) {
        cin >> a[i];
    }

    // Делим массив на две половины, чтобы перебирать их независимо.
    int mid = n / 2;
    vector<long long> left(a.begin(), a.begin() + mid);
    vector<long long> right(a.begin() + mid, a.end());

    // Генерируем все суммы отдельно для каждой половины.
    vector<long long> left_sums = build_sums(left);
    vector<long long> right_sums = build_sums(right);
    sort(right_sums.begin(), right_sums.end());

    long long best = 0;
    for (long long left_sum : left_sums) {
        if (left_sum > limit_sum) continue;

        // Ищем максимальную допустимую правую половину.
        long long need = limit_sum - left_sum;
        auto it = upper_bound(right_sums.begin(), right_sums.end(), need);
        if (it == right_sums.begin()) {
            best = max(best, left_sum);
            continue;
        }
        --it;
        best = max(best, left_sum + *it);
    }

    // Печатаем лучшую допустимую сумму.
    cout << best << '\n';
    return 0;
}